internal/mailer/mailer.go

```diff
@@ -2,16 +2,52 @@
 package mailer
 
 import (
+	"io"
+	"kg/procurement/cmd/utils"
 	"kg/procurement/internal/common/database"
+	"mime/multipart"
 	"time"
 )
 
 type Email struct {
-	From    string
-	To      []string
-	CC      []string
-	Subject string
-	Body    string
+	From        string
+	To          []string
+	CC          []string
+	Subject     string
+	Body        string
+	Attachments []Attachment
+}
+
+type Attachment struct {
+	Filename string
+	Data     []byte
+	MIMEType string
+}
+
+func BulkFromMultipartForm(files []*multipart.FileHeader) ([]Attachment, error) {
+	var attachments []Attachment
+	for _, fileHeader := range files {
+		file, err := fileHeader.Open()
+		if err != nil {
+			utils.Logger.Error(err.Error())
+			return nil, err
+		}
+		defer file.Close()
+
+		fileData, err := io.ReadAll(file)
+		if err != nil {
+			utils.Logger.Error(err.Error())
+			return nil, err
+		}
+
+		attachments = append(attachments, Attachment{
+			Filename: fileHeader.Filename,
+			Data:     fileData,
+			MIMEType: fileHeader.Header.Get("Content-Type"),
+		})
+	}
+
+	return attachments, nil
 }
```

Comment:
I'm new to passing file attachments directly in API requests. Could you enlighten me on how this feature works? Does attachments sent as bytes form and then because of limitation of the size of one request, they split it into multiple parts and in this function you are trying to read all by combining all of the bytes again?

Reply:
Actually I’ve been curious as well about how file attachments pass through API requests. From what I know, the multipart/form-data encoding type splits the attachment into multiple parts and sends them within a single request. The key idea here is that, when dealing with attachments, it's not just the file content that matters, but also the metadata associated with it. For example, file extensions may require custom handling depending on their type, so it’s important to retain that information. To send both the file content and metadata together, multipart/form-data is used.

This separation of metadata and file content makes it easier to handle complex data (like files and form fields) within a single request. The server can read the metadata (such as the filename, MIME type, etc.) from one part, and the actual file content from another part.

That said, do you think there are any performance trade-offs when using this method? For instance, does the overhead of splitting the request into multiple parts or the need for chunking large files affect the upload speed, or is the reduction in memory usage and more efficient file handling worth it in most cases?

Reply:
Great question @valenciusap17
After reading through some articles online, here are some of the things I've learned
- `multipart/form-data` introduces some overhead because each part of the request (metadata and file content) has to be serialized, encoded, and sent separately.
- Here's an interesting one: multipart uploads allow for streaming the file data. The server can process the file as it's being received without needing to load the entire file into memory at once.

But overall multipart provides significant flexibility, memory management, and streaming capabilities that make it a solid choice, especially for larger files.

Other alternatives does exist, for example using direct file storages such as AWS to bypass the server entirely for the file transfer.